### ISSUE TITLE
KQUEUE leak

### ISSUE THREAD

**Issue by spuun** [yxhuvud/nested_scheduler#7]
When the scheduler is stopped, the kqueue isn't cleaned up, so you'll end up with dangling kqueues.

Reproduce with
```crystal
10.times do
  NestedScheduler::ThreadPool.nursery do |pool|
    pool.spawn { sleep 0.1 }
  end
end
sleep
```

Check lsof and you'll see 10 unwanted KQUEUE.

I have no clue on what I'm doing, but I did manage to put this monkey patch together to get rid of the KQUEUEs:

```crystal
abstract class Crystal::EventLoop
  def stop
  end
end

lib LibEvent2
  fun event_base_free(event : EventBase) : Nil
end

struct Crystal::LibEvent::Event::Base
  def free : Nil
    LibEvent2.event_base_free(@base)
  end
end

class Crystal::LibEvent::EventLoop
  def stop
    event_base.free
  end
end
```

and then call `@event_loop.stop` before returning from Scheduler's `run_loop`.

**Comment by yxhuvud** [comment]
Well, no kqueue for me on linux, but the same basic issue is there, a whole bunch of extra files show up in `lsof -p`. Together with `pgrep` it should be possible to figure out some way to write a spec for it. And the suggested solution seems good though it should involve the io_context. 

I'll have some time tomorrow as I have a longer train trip planned, lets see if I can piece it together.

**Comment by spuun** [comment]
I should have mentioned:

> Crystal 1.9.2 (2023-07-19)
> LLVM: 15.0.7
> Default target: aarch64-apple-darwin23.1.0

I was actually playing around with my own "worker thread" thing and had some problems. My google fu led me here and then I tested `nested_scheduler` to see if it had the same issues.

**Comment by yxhuvud** [comment]
Also we seem to leak pipe fds. Closing things has sideeffects. A bit more investigation required.

**Comment by spuun** [comment]
I think it's the fiber channel. I did

```crystal
struct Crystal::FiberChannel
  def close
    # Send zero pointer to signal stop
    @worker_in.write_bytes 0u64
  end

  def receive : Fiber?
    oid = @worker_out.read_bytes(UInt64)
    if oid.zero?
      @worker_in.close
      @worker_out.close
      nil
    else
      Pointer(Fiber).new(oid).as(Fiber)
    end
  end
end
```

which seems to help.

**Comment by yxhuvud** [comment]
It does, but it make the test suite fail for some reason. I'll have to dig into it a bit further when I get back to my normal computer.

**Comment by yxhuvud** [comment]
Fixed in 74094c1 . Some issues seems local to my laptop, so there may be hidden pit traps somewhere. But this particular thing should be fixed.
